File: sea-shell-service-immutable/src/main/java/org/jesperancinha/shell/webflux/immutable/repository/ShellAccountImmutableRepository.java
package org.jesperancinha.shell.webflux.immutable.repository;

import org.jesperancinha.shell.client.accounts.Account;
import org.jesperancinha.shell.client.accounts.AccountsClient;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Mono;

import static reactor.core.scheduler.Schedulers.single;


@Repository
public class ShellAccountImmutableRepository {

    private final AccountsClient accountsClient;

    public ShellAccountImmutableRepository(final AccountsClient accountsClient) {
        this.accountsClient = accountsClient;
    }

    public Mono<Account> findAccountById(String id) {
        return Mono.fromCallable(() -> accountsClient.getAccount(id))
                .subscribeOn(single());
    }

}

File: sea-shell-service-immutable/src/main/kotlin/org/jesperancinha/shell/webflux/immutable/repository/ShellAccountImmutableRepository.kt
package org.jesperancinha.shell.webflux.immutable.repository

import org.jesperancinha.shell.client.accounts.Account
import org.jesperancinha.shell.client.accounts.AccountsClient
import org.springframework.stereotype.Repository
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers

@Repository
class ShellAccountImmutableRepository(private val accountsClient: AccountsClient) {
    fun findAccountById(id: String?): Mono<Account?> {
        return Mono.fromCallable { accountsClient.getAccount(id) }
            .subscribeOn(Schedulers.single())
    }
}